src/libraries/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs

```diff
@@ -320,118 +386,223 @@ internal static string GetTextFromUtf8(ReadOnlySpan<byte> utf8Text)
+        {
+            Debug.Assert(destination.Length >= source.Length);
+
+            int idx = source.IndexOf(JsonConstants.BackSlash);
+            Debug.Assert(idx >= 0);
+
+            bool result = TryUnescape(source, destination, idx, out written, clearBufferOnFailure: false);
+            Debug.Assert(result);
+        }
+
         internal static void Unescape(ReadOnlySpan<byte> source, Span<byte> destination, int idx, out int written)
         {
             Debug.Assert(idx >= 0 && idx < source.Length);
             Debug.Assert(source[idx] == JsonConstants.BackSlash);
             Debug.Assert(destination.Length >= source.Length);
 
-            source.Slice(0, idx).CopyTo(destination);
+            bool result = TryUnescape(source, destination, idx, out written, clearBufferOnFailure: false);
+            Debug.Assert(result);
+        }
+
+        /// <summary>
+        /// Used when writing to buffers not guaranteed to fit the unescaped result.
+        /// </summary>
+        internal static bool TryUnescape(ReadOnlySpan<byte> source, Span<byte> destination, out int written, bool clearBufferOnFailure)
+        {
+            int idx = source.IndexOf(JsonConstants.BackSlash);
+            Debug.Assert(idx >= 0);
+
+            return TryUnescape(source, destination, idx, out written, clearBufferOnFailure);
+        }
+
+        /// <summary>
+        /// Used when writing to buffers not guaranteed to fit the unescaped result.
+        /// </summary>
+        private static bool TryUnescape(ReadOnlySpan<byte> source, Span<byte> destination, int idx, out int written, bool clearBufferOnFailure)
+        {
+            Debug.Assert(idx >= 0 && idx < source.Length);
+            Debug.Assert(source[idx] == JsonConstants.BackSlash);
+
+            if (!source.Slice(0, idx).TryCopyTo(destination))
+            {
+                written = 0;
+                goto DestinationTooShort;
+            }
+
             written = idx;
 
-            for (; idx < source.Length; idx++)
+            while (true)
             {
-                byte currentByte = source[idx];
-                if (currentByte == JsonConstants.BackSlash)
+                Debug.Assert(source[idx] == JsonConstants.BackSlash);
+
+                if (written == destination.Length)
                 {
-                    idx++;
-                    currentByte = source[idx];
+                    goto DestinationTooShort;
+                }
 
-                    if (currentByte == JsonConstants.Quote)
-                    {
-                        destination[written++] = JsonConstants.Quote;
-                    }
-                    else if (currentByte == 'n')
-                    {
-                        destination[written++] = JsonConstants.LineFeed;
-                    }
-                    else if (currentByte == 'r')
+                int currentByte = source[++idx];
+                if (currentByte == JsonConstants.Quote)
```

Comment:
Consider using a switch

Reply:
I initially started refactoring this to a switch, but ultimately ended up reverting it to reduce this PR's diff size.